Create notes front-end with cache invalidation

benbaumann95 opened the pull request.

Commit:
refactor ajax js and add support for post

Commit:
post form for create note

Commit:
stop using FormData#get as it is not supported by ie11

benbaumann95 committed:
initial post

benbaumann95 committed:
refactor cache

benbaumann95 committed:
invalidate cache with post note request

benbaumann95 committed:
update ui for adding note

benbaumann95 committed:
form validation

benbaumann95 committed:
new aws base url has / appended unlike the previous one so it must be taken off in the url variable

benbaumann95 committed:
stop api request cache unless in test environment

benbaumann95 committed:
using fake redis

benbaumann95 committed:
refactor destination.html

benbaumann95 committed:
refactor error summary to incorporate max characters

spannerinworks commented on app/controllers/api/work_orders_controller.rb:
```diff
@@ -6,6 +6,21 @@ def description
+
+      if !params[:note][:text].empty?
+        Hackney::Note.create_work_order_note(reference, params[:note][:text])
+        @notes_and_appointments = @work_order.notes + (@work_order.appointments.nil? ? [] : @work_order.appointments)
```
seems odd that `@notes_and_appointments` is only on 1 side of the if statement

spannerinworks commented on app/controllers/api/work_orders_controller.rb:
```diff
@@ -6,6 +6,21 @@ def description
+    def notes
+      @work_order = Hackney::WorkOrder.find(reference)
+
+      if !params[:note][:text].empty?
```
if you're checking empty here you might as well check for 2000. use `present?` instead of `!x.empty?` as it removes the negation and handles nil and blank strings.

spannerinworks replied:
if there is an error here we should render some of message to the user

benbaumann95 committed:
abstract get notes and appointments logic and add max character validation

benbaumann95 committed:
render 400 error if note text is not between 1 and 2000 characters

benbaumann95 committed:
remove need for base url in cache key

This is because the base url may have a different format i.e. with a trailing "/" or without one which would make the cache invalidation not work
